source/extensions/filters/http/custom_response/policy.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/stream_info/stream_info.h"
+
+#include "source/common/matcher/matcher.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CustomResponse {
+
+class CustomResponseFilter;
+
+// Base class for custom response policies.
+class Policy : public std::enable_shared_from_this<Policy>, public StreamInfo::FilterState::Object {
+public:
+  ~Policy() override = default;
+
+  virtual Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap&, bool,
+                                                  CustomResponseFilter&) const PURE;
+
+protected:
+  Policy() = default;
+};
+
+using PolicySharedPtr = std::shared_ptr<Policy>;
```

Comment:
Do you need Policy based objects to be mutable after they were created? If not, it would be better to make shared pointer of `const Policy`. This is again to make it easier to reason about config immutability.

Reply:
Done.